Topic refactor cuda hip atomics

psychocoderHPC opened the pull request:
In https://github.com/alpaka-group/alpaka/pull/1767 I already enabled the usage of building atomicAdd for double precision.
Nevertheless, we are emulating in alpaka often atomics because it is a nightmare to maintain the list of supported atomic functions for HIP and CUDA. The list of supported types for different atomics differs between both models.

This PR introduces a "simple" way to let the compiler decide if a built-in function is available and use these. If not we will emulate the required atomics by using `atomicCas`.

- enable atomicCas for floating point data types
- emulate atomics if no builtin is available
- alpaka::AtomicCas add floating point support
- refactor atomic unit tests (remove many exceptions for CUDA/HIP)
- use builtin atomics for `hierarchy::Threads` -> mapped to `atomicXXX_block()`

psychocoderHPC committed:
refactor HIP/CUDA atomic implementations

- check for builtin atomics and use these if available
- emulate atomics which are not provided as builtin via atomicCas
- emulate atomicCas for floating point data

psychocoderHPC committed:
`alpaka::AtomicCas` add floating point support

psychocoderHPC committed:
refactore atomic unit tests

Unify CPU and HIP/CUDA atomic tests.
For CUDA run tests only for 32bit and 64bit data types.

SimeonEhrig commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+    //
+    //  Atomics can be used in the hierarchy level grids, blocks and threads.
+    //  Atomics are not guaranteed to be save between devices.
+    class AtomicUniformCudaHipBuiltIn
```
Does it also works, if the CUDA and the HIP are enabled at the same time?

psychocoderHPC replied:
As I know. You can not enable CUDA and HIP at the same time.

SimeonEhrig replied:
Really, I thought Andrea implemented it? Or is it permitted by the CMake?

psychocoderHPC replied:
As I know he is compiling a library for CUDA and a library for HIP and linking both together.
Both libs will use a different compiler.

@fwyzard Could you check if this PR is working for you. I do not exactly know your workflows.

psychocoderHPC replied:
I think I remember that we introduced the device API struct to avoid naming conflicts. Nevertheless, these changes are fully on the device side and should not create symbols on the host side which can conflict.

fwyzard replied:
Sorry, I have been doing night shifts this week, and had little time for the rest :-/

I'll try to have a look ASAP.

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+    //! The GPU CUDA/HIP accelerator atomic ops.
+    //
+    //  Atomics can be used in the hierarchy level grids, blocks and threads.
+    //  Atomics are not guaranteed to be save between devices.
```
```suggestion
    //  Atomics are not guaranteed to be safe between devices.
```

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+#            error If ALPAKA_ACC_GPU_HIP_ENABLED is set, the compiler has to support HIP!
+#        endif
+
+//! Inline namespace is required to check the existenz of function in global namespace via SFINAE.
```
```suggestion
//! These types must be in the global namespace for checking existence of respective functions in global namespace via SFINAE, so we use inline namespace.
```

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+    //! Provide an interface to builtin atomic functions.
+    //
+    // To check for the existence of builtin functions located in the global namespace :: directly.
+    // This is forbidden by C++.
```
I do not understand this line. Did you mean smth like
```suggestion
    // This would not be possible without having these types in global namespace.
```
?

psychocoderHPC replied:
Yes! Your suggestion is more clear!

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+    // This is forbidden by C++.
+    // If the functor is inheriting from std::false_type an signature is explicitly not available. This can be used to
+    // explicitly disable builtin function in case the builtin is broken.
+    // If the functor is inheriting from std::true_type a spezialization must implement one of the following
```
```suggestion
    // If the functor is inheriting from std::true_type a specialization must implement one of the following
```

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+    // \code{.cpp}
+    //    __device__ static T atomic( T* add, T value);
+    //    __device__ static T atomic( T* add, T compare, T value);
+    // \endcode
+    template<typename TOp, typename T, typename THierarchy, typename TSfinae = void, typename TDefer = void>
```
I think doxygen comment on template parameters would be helpful here. Perhaps some parts of the comment above could be moved there.

psychocoderHPC replied:
I expect the comment would be as useful as the comment from atomicOp. Therefore I decided to have no comment.

```
    //! Executes the given operation atomically.
    //!
    //! \tparam TOp The operation type.
    //! \tparam T The value type.
    //! \tparam TAtomic The atomic implementation type.
    //! \param addr The value to change atomically.
    //! \param value The value used in the atomic operation.
    //! \param atomic The atomic implementation.
    ALPAKA_NO_HOST_ACC_WARNING
    template<typename TOp, typename TAtomic, typename T, typename THierarchy = hierarchy::Grids>
  ALPAKA_FN_HOST_ACC auto atomicOp(
```

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+        typename std::void_t<decltype(atomicCAS(std::declval<T*>(), std::declval<T>(), std::declval<T>()))>>
+        : std::true_type
+    {
+        __device__ static T atomic(T* add, T compare, T value)
```
Why did you choose name `atomic()` and not `operator()` like math traits do? Not against that, just wondering about the reasoning

psychocoderHPC replied:
because the method is static same as the methods in `AtomicOp`

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+        alpaka::AtomicCas,
+        T,
+        THierarchy,
+        typename std::void_t<decltype(atomicCAS(std::declval<T*>(), std::declval<T>(), std::declval<T>()))>>
```
I believe on device side we must use `alpaka::core::declval` and not `std::declval`, please see the comment there.

psychocoderHPC committed:
workaround for clang 9 with cuda 9.2

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+        alpaka::AtomicMin,
+        float,
+        THierarchy,
+        typename std::void_t<
+            decltype(atomicMin(std::declval<float*>(), std::declval<std::enable_if_t<sizeof(THierarchy), float>>()))>>
```
I am not sure this part is needed. If we jsut want to hard-disable for HIP and given type, shouldn't we do it for any `TSfinae`?

Same point for specializations below.

psychocoderHPC replied:
Yes you are right it is not required!

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHip.hpp:
```diff
@@ -0,0 +1,474 @@
+    // If the functor is inheriting from std::false_type an signature is explicitly not available. This can be used to
+    // explicitly disable builtin function in case the builtin is broken.
+    // If the functor is inheriting from std::true_type a spezialization must implement one of the following
+    // interfaces.
```
(Optional) we could clarify that all but CAS follow the first interface and CAS follows the second - there is no optionality as now suggested by the comment.

psychocoderHPC replied:
I addressed it in the code snipped

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp:
```diff
@@ -37,998 +30,301 @@ namespace alpaka
-                if constexpr(std::numeric_limits<unsigned int>::max() == std::numeric_limits<unsigned long>::max())
-                    return ::atomicAdd(reinterpret_cast<unsigned int*>(addr), static_cast<unsigned int>(value));
-                else // LP64
+                //! reinterprets an adress to an 32bit value for atomicCas emulation usage
```
```suggestion
                //! reinterprets an address as a 32bit value for atomicCas emulation usage
```

sbastrakov replied:
Also for later usage of the same phrase

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp:
```diff
@@ -37,998 +30,301 @@ namespace alpaka
+                //! reinterprets an adress to an 32bit value for atomicCas emulation usage
+                template<typename TAddressType>
+                __device__ static auto reinterpretAddress(TAddressType* address)
+                    -> std::enable_if_t<sizeof(TAddressType) == 4u, unsigned int*>
```
Should a fixed-width type be used instead of `unsigned int`? Then the 32-bit requirement is also more obvious. However I am not sure if it bodes well with CUDA and HIP.

psychocoderHPC replied:
No, we need to use a type where we know that the signature fits CUDA/HIP. If we use `std::uint32_t` it can link to `unsigned long int` or `unsigned int` depending on the platform.

sbastrakov replied:
Okay, but then we should also check that `sizeof(unsigned int) == 4` ?

sbastrakov replied:
Discussed offline: maybe it's not worth as nothing would work with CUDA/HIP if `unsigned int` is not 4 bytes.

psychocoderHPC replied:
The CUDA and HIP ecosystem require that `sizeof(int)==4`. Seeing it from the C++ standard view you are right but we do not guard it in many other places. 
We could at somewhere a global check for this condition but in reality, we will not run on systems where this is not fulfilled.

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp:
```diff
@@ -37,998 +30,301 @@ namespace alpaka
-                int old(*addr);
-                int assumed, eval;
-                do
+                //! reinterprets an value to be usable for the atomicCAS emulation
```
```suggestion
                //! reinterprets a value to be usable for the atomicCAS emulation
```

sbastrakov commented on include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp:
```diff
@@ -37,998 +30,301 @@ namespace alpaka
+                        using Cas = alpaka::trait::
+                            AtomicOp<alpaka::AtomicCas, alpaka::AtomicUniformCudaHipBuiltIn, EmulatedType, THierarchy>;
+                        old = Cas::atomicOp(ctx, addressAsIntegralType, assumed, reinterpretValue(v));
+                        // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
```
Do we have this case covered in tests? It should have been even before this PR, but since you are working on this part

psychocoderHPC replied:
No, we do not cover this case.

psychocoderHPC replied:
Our checks for atomicDec also do not test the flip over. https://github.com/alpaka-group/alpaka/issues/1769

psychocoderHPC committed:
disable OpenACC float comparisn warning

```
alpaka/atomic/AtomicOaccExtended.hpp:209:36: error: comparing floating-point with '==' or '!=' is unsafe [-Werror=float-equal]
```

psychocoderHPC committed:
OpenACC: test only 32bit and 64bit atomics

psychocoderHPC committed:
HIP: usa atomic load within atomicCas emulation

psychocoderHPC committed:
HIP: workaround atomicMax and atomicMin

For architectures where __hip_atomic_compare_exchange_strong is false
HIP is emulating atomics with `atomicCAS` but with a broken
implementation.
